File: ts/src/body.ts
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci
/**
 * Parameter validation and the generic body serialiser, driven by the transaction descriptions
 * of spec/transactions (the encodings of index.json). Fields marked `computed` are produced by
 * the hooks in custom.ts.
 */
import { parseAddress, parseKey32 } from "./address.js";
import { keyPairFromSeed, KeyPair } from "./keys.js";
import { usage } from "./errors.js";
import { ByteWriter, hexToBytes, isHex, utf8 } from "./util/bytes.js";
import { FieldDef, ParamDef, TxDef } from "./spec.js";

const INT_LIMITS: Record<string, [bigint, bigint]> = {
  int64: [-(1n << 63n), (1n << 63n) - 1n], uint64: [0n, (1n << 64n) - 1n], uint32: [0n, (1n << 32n) - 1n], uint8: [0n, 255n],
};

export function parseInteger(value: string, kind: string, name: string): bigint {
  if (!/^-?\d+$/.test(value.trim())) throw usage(`${name} must be a whole number, got "${value}"`);
  const v = BigInt(value.trim());
  const lim = INT_LIMITS[kind] ?? INT_LIMITS.int64;
  if (v < lim[0] || v > lim[1]) throw usage(`${name} is out of range for ${kind}`);
  return v;
}

/** Check one parameter value against its kind; returns the value to keep (trimmed where it matters). */
export function validateParam(p: ParamDef, value: string): string {
  switch (p.kind) {
    case "privkey":
      if (!isHex(value, 64)) throw usage(`${p.name} must be 64 hex characters (a 32-byte private key)`);
      return value;
    case "address":
      try { parseAddress(value); } catch (e) { throw usage(`invalid ${p.name}: ${(e as Error).message}`); }
      return value;
    case "address_list":
      for (const a of splitList(value)) { try { parseAddress(a); } catch (e) { throw usage(`invalid ${p.name} entry ${a}: ${(e as Error).message}`); } }
      return value;
    case "key":
      try { parseKey32(value); } catch (e) { throw usage(`invalid ${p.name}: ${(e as Error).message}`); }
      return value;
    case "int64": case "uint64": case "uint32": case "uint8": {
      const n = parseInteger(value, p.kind, p.name);
      if ((p.min !== undefined && n < BigInt(p.min)) || (p.max !== undefined && n > BigInt(p.max)))
        throw usage(`${p.name} must be between ${p.min ?? "-inf"} and ${p.max ?? "inf"}`);
      return value.trim();
    }
    case "hex32":
      if (!isHex(value, 64)) throw usage(`${p.name} must be 64 hex characters (32 bytes)`);
      return value;
    case "hexbytes":
      if (!isHex(value)) throw usage(`${p.name} must be hex`);
      return value;
    case "string": case "file":
      return value;
    default:
      return value;
  }
}

export function splitList(s: string): string[] {
  return s.split(",").map((x) => x.trim()).filter((x) => x.length > 0);
}

export interface BodyContext {
  /** The signing key pair (for `sender_address`). */
  sender: KeyPair;
  /** Bytes of `file` parameters, by parameter name, as hex. */
  files?: Record<string, string>;
  /** Values produced by custom hooks, by field name. */
  computed?: Record<string, Uint8Array>;
}

function conditionHolds(when: string, params: Record<string, string>): boolean {
  const m = /^(\w+) != (0|'')$/.exec(when);
  if (!m) throw new Error("unsupported condition " + when);
  const v = params[m[1]] ?? "";
  return m[2] === "0" ? BigInt(v || "0") !== 0n : v !== "";
}

/** Serialise one body field. */
export function encodeField(f: FieldDef, params: Record<string, string>, ctx: BodyContext): Uint8Array {
  const w = new ByteWriter();
  if (ctx.computed && f.name in ctx.computed) return ctx.computed[f.name];
  let value = params[f.from] ?? "";
  if (f.when_zero && (value === "" || BigInt(value) <= 0n)) value = params[f.when_zero] ?? "0";
  switch (f.encoding) {
    case "u8": return w.u8(Number(parseInteger(value, "uint8", f.name))).finish();
    case "u16le": return w.u16(Number(parseInteger(value, "uint32", f.name))).finish();
    case "u32le": return w.u32(Number(parseInteger(value, "uint32", f.name))).finish();
    case "u64le": return w.i64(parseInteger(value, "int64", f.name)).finish();
    case "hex": {
      const b = hexToBytes(value);
      if (f.size !== undefined && b.length !== f.size) throw usage(`${f.from} must be ${f.size} bytes (${2 * f.size} hex)`);
      return b;
    }
    case "hex16": { const b = hexToBytes(value); return w.u16(b.length).bytes(b).finish(); }
    case "bytes32": {
      const b = hexToBytes(ctx.files && f.from in ctx.files ? ctx.files[f.from] : value);
      return w.u32(b.length).bytes(b).finish();
    }
    case "str16": { const b = utf8(value); return w.u16(b.length).bytes(b).finish(); }
    case "str32": { const b = utf8(value); return w.u32(b.length).bytes(b).finish(); }
    case "address": return parseAddress(value).bytes;
    case "address_list": { for (const a of splitList(value)) w.bytes(parseAddress(a).bytes); return w.finish(); }
    case "address_list8": {
      const items = splitList(value);
      if (items.length > 255) throw usage(`${f.from}: at most 255 entries`);
      w.u8(items.length);
      for (const a of items) w.bytes(parseAddress(a).bytes);
      return w.finish();
    }
    case "sender_address": return ctx.sender.accountBytes;
    case "pubkey_of_key": {
      if (!isHex(value, 64)) throw usage(`${f.from} must be 64 hex characters (a 32-byte private key)`);
      return keyPairFromSeed(value).publicKey;
    }
    case "key32": return parseKey32(value);
    case "literal": return hexToBytes(f.value ?? "");
    case "custom": throw new Error(`field ${f.name} of ${f.from} needs a custom hook`);
    default: throw new Error("unknown encoding " + f.encoding);
  }
}

/** The whole body of a non-custom transaction (custom hooks may pre-fill `ctx.computed`). */
export function buildBody(def: TxDef, params: Record<string, string>, ctx: BodyContext): Uint8Array {
  const w = new ByteWriter();
  for (const f of def.body) {
    if (f.when && !conditionHolds(f.when, params)) continue;
    w.bytes(encodeField(f, params, ctx));
  }
  return w.finish();
}

File: go/zbc/body.go
// SPDX-License-Identifier: MIT
// Copyright (c) 2024-2026 ZooBC Foundation and Roberto Capodieci

package zbc

import (
	"encoding/hex"
	"fmt"
	"regexp"
	"strconv"
	"strings"
)

var intRe = regexp.MustCompile(`^-?\d+$`)

// ParseInteger reads a decimal parameter of the given kind (int64, uint64, uint32, uint8).
func ParseInteger(value, kind, name string) (int64, error) {
	v := strings.TrimSpace(value)
	if !intRe.MatchString(v) {
		return 0, Usage("%s must be a whole number, got \"%s\"", name, value)
	}
	switch kind {
	case "uint64":
		u, err := strconv.ParseUint(v, 10, 64)
		if err != nil {
			return 0, Usage("%s is out of range for %s", name, kind)
		}
		return int64(u), nil
	case "uint32", "uint8":
		u, err := strconv.ParseUint(v, 10, map[string]int{"uint32": 32, "uint8": 8}[kind])
		if err != nil {
			return 0, Usage("%s is out of range for %s", name, kind)
		}
		return int64(u), nil
	default:
		n, err := strconv.ParseInt(v, 10, 64)
		if err != nil {
			return 0, Usage("%s is out of range for %s", name, kind)
		}
		return n, nil
	}
}

// SplitList splits a comma-separated parameter, dropping blanks.
func SplitList(s string) []string {
	var out []string
	for _, x := range strings.Split(s, ",") {
		if t := strings.TrimSpace(x); t != "" {
			out = append(out, t)
		}
	}
	return out
}

// ValidateParam checks one value against its parameter kind; returns the value to keep.
func ValidateParam(p ParamDef, value string) (string, error) {
	switch p.Kind {
	case "privkey":
		if !IsHex(value, 64) {
			return "", Usage("%s must be 64 hex characters (a 32-byte private key)", p.Name)
		}
	case "address":
		if _, err := ParseAddress(value, ""); err != nil {
			return "", Usage("invalid %s: %v", p.Name, err)
		}
	case "address_list":
		for _, a := range SplitList(value) {
			if _, err := ParseAddress(a, ""); err != nil {
				return "", Usage("invalid %s entry %s: %v", p.Name, a, err)
			}
		}
	case "key":
		if _, err := ParseKey32(value); err != nil {
			return "", Usage("invalid %s: %v", p.Name, err)
		}
	case "int64", "uint64", "uint32", "uint8":
		n, err := ParseInteger(value, p.Kind, p.Name)
		if err != nil {
			return "", err
		}
		if (p.Min != nil && n < *p.Min) || (p.Max != nil && n > *p.Max) {
			lo, hi := "-inf", "inf"
			if p.Min != nil {
				lo = strconv.FormatInt(*p.Min, 10)
			}
			if p.Max != nil {
				hi = strconv.FormatInt(*p.Max, 10)
			}
			return "", Usage("%s must be between %s and %s", p.Name, lo, hi)
		}
		return strings.TrimSpace(value), nil
	case "hex32":
		if !IsHex(value, 64) {
			return "", Usage("%s must be 64 hex characters (32 bytes)", p.Name)
		}
	case "hexbytes":
		if !IsHex(value, 0) {
			return "", Usage("%s must be hex", p.Name)
		}
	}
	return value, nil
}

// BodyContext carries what the generic serialiser cannot read from the parameters.
type BodyContext struct {
	Sender   KeyPair
	Files    map[string][]byte // bytes of `file` parameters, by parameter name
	Computed map[string][]byte // values produced by custom hooks, by field name
}

var whenRe = regexp.MustCompile(`^(\w+) != (0|'')$`)

func conditionHolds(when string, params map[string]string) (bool, error) {
	m := whenRe.FindStringSubmatch(when)
	if m == nil {
		return false, fmt.Errorf("unsupported condition %s", when)
	}
	v := params[m[1]]
	if m[2] == "0" {
		if v == "" {
			return false, nil
		}
		n, err := strconv.ParseInt(v, 10, 64)
		return err == nil && n != 0, nil
	}
	return v != "", nil
}

// EncodeField serialises one body field.
func EncodeField(f FieldDef, params map[string]string, ctx *BodyContext) ([]byte, error) {
	if ctx.Computed != nil {
		if v, ok := ctx.Computed[f.Name]; ok {
			return v, nil
		}
	}
	value := params[f.From]
	if f.WhenZero != "" {
		if n, err := strconv.ParseInt(value, 10, 64); value == "" || (err == nil && n <= 0) {
			value = params[f.WhenZero]
			if value == "" {
				value = "0"
			}
		}
	}
	w := NewWriter()
	switch f.Encoding {
	case "u8":
		n, err := ParseInteger(value, "uint8", f.Name)
		if err != nil {
			return nil, err
		}
		return w.U8(byte(n)).Finish(), nil
	case "u16le":
		n, err := ParseInteger(value, "uint32", f.Name)
		if err != nil {
			return nil, err
		}
		return w.U16(uint16(n)).Finish(), nil
	case "u32le":
		n, err := ParseInteger(value, "uint32", f.Name)
		if err != nil {
			return nil, err
		}
		return w.U32(uint32(n)).Finish(), nil
	case "u64le":
		n, err := ParseInteger(value, "int64", f.Name)
		if err != nil {
			return nil, err
		}
		return w.I64(n).Finish(), nil
	case "hex":
		b, err := hex.DecodeString(value)
		if err != nil {
			return nil, Usage("%s must be hex", f.From)
		}
		if f.Size > 0 && len(b) != f.Size {
			return nil, Usage("%s must be %d bytes (%d hex)", f.From, f.Size, 2*f.Size)
		}
		return b, nil
	case "hex16":
		b, err := hex.DecodeString(value)
		if err != nil {
			return nil, Usage("%s must be hex", f.From)
		}
		return w.U16(uint16(len(b))).Bytes(b).Finish(), nil
	case "bytes32":
		b, ok := ctx.Files[f.From]
		if !ok {
			var err error
			if b, err = hex.DecodeString(value); err != nil {
				return nil, Usage("%s must be hex", f.From)
			}
		}
		return w.U32(uint32(len(b))).Bytes(b).Finish(), nil
	case "str16":
		return w.U16(uint16(len(value))).Bytes([]byte(value)).Finish(), nil
	case "str32":
		return w.U32(uint32(len(value))).Bytes([]byte(value)).Finish(), nil
	case "address":
		a, err := ParseAddress(value, "")
		if err != nil {
			return nil, Usage("invalid %s: %v", f.From, err)
		}
		return a.Bytes(), nil
	case "address_list", "address_list8":
		items := SplitList(value)
		if f.Encoding == "address_list8" {
			if len(items) > 255 {
				return nil, Usage("%s: at most 255 entries", f.From)
			}
			w.U8(byte(len(items)))
		}
		for _, x := range items {
			a, err := ParseAddress(x, "")
			if err != nil {
				return nil, Usage("invalid %s entry %s: %v", f.From, x, err)
			}
			w.Bytes(a.Bytes())
		}
		return w.Finish(), nil
	case "sender_address":
		return ctx.Sender.AccountBytes(), nil
	case "pubkey_of_key":
		kp, err := KeyPairFromHex(value)
		if err != nil {
			return nil, Usage("%s must be 64 hex characters (a 32-byte private key)", f.From)
		}
		return kp.PublicKey, nil
	case "key32":
		b, err := ParseKey32(value)
		if err != nil {
			return nil, Usage("invalid %s: %v", f.From, err)
		}
		return b, nil
	case "literal":
		b, err := hex.DecodeString(f.Value)
		return b, err
	case "custom":
		return nil, fmt.Errorf("field %s needs a custom hook", f.Name)
	}
	return nil, fmt.Errorf("unknown encoding %s", f.Encoding)
}

// BuildBody serialises the whole body of a non-custom transaction.
func BuildBody(def *TxDef, params map[string]string, ctx *BodyContext) ([]byte, error) {
	w := NewWriter()
	for _, f := range def.Body {
		if f.When != "" {
			ok, err := conditionHolds(f.When, params)
			if err != nil {
				return nil, err
			}
			if !ok {
				continue
			}
		}
		b, err := EncodeField(f, params, ctx)
		if err != nil {
			return nil, err
		}
		w.Bytes(b)
	}
	return w.Finish(), nil
}
